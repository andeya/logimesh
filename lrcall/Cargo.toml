[package]
name = "lrcall"
description = "lrcall is a Rust communication framework that is compatible with local invocations and RPC."
readme = "README.md"
keywords = ["rpc", "network", "server", "api", "microservices"]
categories = ["asynchronous", "network-programming"]
rust-version.workspace = true
authors.workspace = true
license.workspace = true
homepage.workspace = true
repository.workspace = true
version.workspace = true
edition.workspace = true

[lib]
test = true            # Is tested by default.
doctest = true         # Documentation examples are tested by default.
bench = true           # Is benchmarked by default.
doc = true             # Is documented by default.
proc-macro = false     # Set to `true` for a proc-macro library.
harness = true         # Use libtest harness.
crate-type = ["lib"]   # The crate types to generate.
required-features = [] # Features required to build this target (N/A for lib).

[dependencies]

[dev-dependencies]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
