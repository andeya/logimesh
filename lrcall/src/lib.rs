#[cfg(feature = "serde1")]
pub use lrcall_plugins::derive_serde;
pub fn add(left: u64, right: u64) -> u64 {
    left + right
}

/// The main macro that creates RPC services.
///
/// Rpc methods are specified, mirroring trait syntax:
///
/// ```
/// #[lrcall::service]
/// trait Service {
/// /// Say hello
/// async fn hello(name: String) -> String;
/// }
/// ```
///
/// Attributes can be attached to each rpc. These attributes
/// will then be attached to the generated service traits'
/// corresponding `fn`s, as well as to the client stubs' RPCs.
///
/// The following items are expanded in the enclosing module:
///
/// * `trait Service` -- defines the RPC service.
///   * `fn serve` -- turns a service impl into a request handler.
/// * `Client` -- a client stub with a fn for each RPC.
///   * `fn new_stub` -- creates a new Client stub.
pub use lrcall_plugins::service;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_works() {
        let result = add(2, 2);
        assert_eq!(result, 4);
    }
}