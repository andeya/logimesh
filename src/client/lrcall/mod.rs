// Copyright Andeya Lee 2024
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.
//! Provides a Stub trait, implemented by types that can call remote services.

pub use lrcall::*;

mod lrcall;
