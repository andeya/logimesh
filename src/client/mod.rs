//! Provides a client that connects to a server and sends multiplexed requests.

pub mod stub;
pub use ::tarpc::client::{new, Channel, Config, NewClient, RequestDispatch, RpcError};
