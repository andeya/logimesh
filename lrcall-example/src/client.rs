// Modifications Copyright Andeya Lee 2024
// Based on original source code from Google LLC licensed under MIT
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

use clap::Parser;
use lrcall::tokio_serde::formats::Json;
use lrcall::{client, context};
use service::{init_tracing, HelloService, WorldChannel, WorldClient};
use std::net::SocketAddr;
use std::time::Duration;
use tokio::time::sleep;
use tracing::Instrument;

#[derive(Parser)]
struct Flags {
    /// Sets the server address to connect to.
    #[clap(long)]
    server_addr: SocketAddr,
    /// Sets the name to say hello to.
    #[clap(long)]
    name: String,
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let flags = Flags::parse();
    init_tracing("Tarpc Example Client")?;

    let mut transport = lrcall::serde_transport::tcp::connect(flags.server_addr, Json::default);
    transport.config_mut().max_frame_length(usize::MAX);

    // WorldClient is generated by the service attribute. It has a constructor `new` that takes a
    // config and any Transport as input.
    let client = WorldClient::full_client(HelloService, WorldChannel::spawn(client::Config::default(), transport.await?));

    let hello = async move {
        // Send the request twice, just to be safe! ;)
        tokio::select! {
            hello1 = client.hello(context::local_current(), format!("{}1", flags.name)) => { hello1 }
            hello2 = client.hello(context::rpc_current(), format!("{}2", flags.name)) => { hello2 }
        }
    }
    .instrument(tracing::info_span!("Two Hellos"))
    .await;

    match hello {
        Ok(hello) => tracing::info!("{hello:?}"),
        Err(e) => tracing::warn!("{:?}", anyhow::Error::from(e)),
    }

    // Let the background span processor finish.
    sleep(Duration::from_micros(10)).await;
    opentelemetry::global::shutdown_tracer_provider();

    Ok(())
}
