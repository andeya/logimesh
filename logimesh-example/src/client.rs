// Modifications Copyright Andeya Lee 2024
// Based on original source code from Google LLC licensed under MIT
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

use clap::Parser;
use logimesh::client::stub::LRConfig;
use logimesh::client::stub::TransportCodec::Json;
use logimesh::context;
use logimesh::discover::{service_lookup, ServiceInfo};
use service::{init_tracing, CompHello, World as _};
use std::net::SocketAddr;
use std::sync::Arc;
use std::time::Duration;
use tokio::time::sleep;
use tracing::Instrument;
#[derive(Parser)]
struct Flags {
    /// Sets the server address to connect to.
    #[clap(long)]
    server_addr: SocketAddr,
    /// Sets the name to say hello to.
    #[clap(long)]
    name: String,
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let flags = Flags::parse();
    init_tracing("Tarpc Example Client")?;

    let client = CompHello
        .logimesh_client(
            LRConfig::new(
                "p.s.m".into(),
                service_lookup(|service_name| Ok(Arc::new(ServiceInfo::new(service_name.into(), vec![flags.server_addr.to_string()])))),
            )
            .with_transport_codec(Json),
        )
        .await?;

    let hello = async move {
        // Send the request twice, just to be safe! ;)
        tokio::select! {
            hello1 = client.hello(context::current(), format!("{}1", flags.name)) => { hello1 }
            hello2 = client.hello(context::current(), format!("{}2", flags.name)) => { hello2 }
        }
    }
    .instrument(tracing::info_span!("Two Hellos"))
    .await;

    match hello {
        Ok(hello) => tracing::info!("{hello:?}"),
        Err(e) => tracing::warn!("{:?}", anyhow::Error::from(e)),
    }

    // Let the background span processor finish.
    sleep(Duration::from_micros(10)).await;
    opentelemetry::global::shutdown_tracer_provider();

    Ok(())
}
