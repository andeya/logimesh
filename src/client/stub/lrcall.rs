// Copyright Andeya Lee 2024
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.
//! A client stbu that supports both local calls and remote calls.
#![allow(dead_code)]

use std::cell::Cell;
use std::sync::Arc;

use crate::client::stub::config::TransportCodec;
use crate::client::stub::{formats, Config, Stub};
use crate::client::{Channel, RpcError};
use crate::discover::ServiceInfo;
use crate::serde_transport::tcp;
use crate::{context, discover, server};

/// Create a builder for a client stub that supports both local and remote calls.
pub struct NewLRCall<Serve, ServiceLookup, RetryFn>
where
    Serve: server::Serve,
{
    lrcall: LRCall<Serve, ServiceLookup, RetryFn>,
}

impl<Serve, ServiceLookup, RetryFn> NewLRCall<Serve, ServiceLookup, RetryFn>
where
    Serve: server::Serve + Clone,
    Serve::Req: crate::serde::Serialize + Send + 'static,
    Serve::Resp: for<'de> crate::serde::Deserialize<'de> + Send + 'static,
    ServiceLookup: discover::ServiceLookup,
    RetryFn: Fn(&Result<Serve::Resp, RpcError>, u32) -> bool,
{
    /// Return a new client stbu that supports both local calls and remote calls.
    pub fn new(serve: Serve, config: Config<ServiceLookup>, retry_fn: RetryFn) -> Self {
        let mut stub_config = tarpc::client::Config::default();
        stub_config.max_in_flight_requests = config.max_in_flight_requests;
        stub_config.pending_request_buffer = config.pending_request_buffer;
        Self {
            lrcall: LRCall {
                serve,
                config,
                stub_config,
                retry_fn,
                rpc_channels: Cell::new(vec![]),
                warm_up_error: None,
            },
        }
    }

    /// Spawn a client stbu that supports both local calls and remote calls.
    pub async fn spawn(self) -> Result<LRCall<Serve, ServiceLookup, RetryFn>, RpcError> {
        self.lrcall.warm_up().await
    }
}

/// A client stbu that supports both local calls and remote calls.
pub struct LRCall<Serve, ServiceLookup, RetryFn>
where
    Serve: server::Serve,
{
    config: Config<ServiceLookup>,
    stub_config: tarpc::client::Config,
    serve: Serve,
    retry_fn: RetryFn,
    rpc_channels: Cell<Vec<ChannelWithInfo<Serve>>>,
    warm_up_error: Option<anyhow::Error>,
}

impl<Serve, ServiceLookup, RetryFn> LRCall<Serve, ServiceLookup, RetryFn>
where
    Serve: server::Serve + Clone,
    Serve::Req: crate::serde::Serialize + Send + 'static,
    Serve::Resp: for<'de> crate::serde::Deserialize<'de> + Send + 'static,
    ServiceLookup: discover::ServiceLookup,
    RetryFn: Fn(&Result<Serve::Resp, RpcError>, u32) -> bool,
{
    /// Warm up client stubs.
    async fn warm_up(self) -> Result<Self, RpcError> {
        self.init_lb_stub().await?;
        Ok(self)
    }
    fn lookup_service(&self) -> anyhow::Result<Arc<ServiceInfo>> {
        self.config.service_lookup.lookup_service(self.config.service_name.as_str())
    }
    async fn init_lb_stub(&self) -> Result<(), RpcError> {
        // TODO
        let service_info = self.lookup_service();
        match service_info {
            Ok(service_info) => {
                self.rpc_channels.set(self.new_channels_with_info(service_info).await?);
            },
            Err(e) => {
                // TODO
                println!("{:?}", e)
            },
        }
        Ok(())
    }
    async fn new_channels_with_info(&self, service_info: Arc<ServiceInfo>) -> Result<Vec<ChannelWithInfo<Serve>>, RpcError> {
        let mut stub_list = vec![];
        match service_info.call_type {
            discover::CallType::Local => {},
            discover::CallType::Remote => {
                for address in service_info.addresses.clone() {
                    match ChannelWithInfo::new(address, self.config.transport_codec.clone(), self.stub_config.clone()).await {
                        core::result::Result::Ok(channel) => {
                            stub_list.push(channel);
                        },
                        Err(e) => {
                            // TODO
                            println!("{:?}", e)
                        },
                    }
                }
            },
        }
        Ok(stub_list)
    }
}

impl<Serve, ServiceLookup, RetryFn> Stub for LRCall<Serve, ServiceLookup, RetryFn>
where
    Serve: server::Serve + Clone,
    ServiceLookup: discover::ServiceLookup,
    RetryFn: Fn(&Result<Serve::Resp, RpcError>, u32) -> bool,
{
    type Req = Serve::Req;
    type Resp = Serve::Resp;

    async fn call(&self, _: context::Context, _request: Self::Req) -> Result<Self::Resp, RpcError> {
        todo!()
    }
}

struct ChannelWithInfo<Serve: server::Serve> {
    address: String,
    transport_codec: TransportCodec,
    stub_config: tarpc::client::Config,
    channel: Channel<Serve::Req, Serve::Resp>,
    is_shutdown: Cell<bool>,
}

impl<Serve: server::Serve> crate::client::stub::Stub for ChannelWithInfo<Serve> {
    type Req = Serve::Req;
    type Resp = Serve::Resp;

    async fn call(&self, ctx: context::Context, request: Self::Req) -> Result<Self::Resp, RpcError> {
        let res = self.channel.call(ctx, request).await;
        if let Err(RpcError::Shutdown) = res {
            self.is_shutdown.set(true);
        }
        res
    }
}

impl<Serve: server::Serve> ChannelWithInfo<Serve>
where
    Serve: server::Serve + Clone,
    Serve::Req: crate::serde::Serialize + Send + 'static,
    Serve::Resp: for<'de> crate::serde::Deserialize<'de> + Send + 'static,
{
    async fn new(address: String, transport_codec: TransportCodec, stub_config: tarpc::client::Config) -> Result<Self, anyhow::Error> {
        let channe = Self::new_channel(transport_codec.clone(), stub_config.clone(), address.as_str()).await?;
        Ok(Self {
            address,
            transport_codec,
            stub_config,
            channel: channe,
            is_shutdown: Cell::new(false),
        })
    }
    async fn new_channel(transport_codec: TransportCodec, stub_config: tarpc::client::Config, address: &str) -> Result<Channel<Serve::Req, Serve::Resp>, anyhow::Error> {
        match transport_codec {
            TransportCodec::Bincode => {
                // Bincode codec using [bincode](https://docs.rs/bincode) crate.
                let mut conn = tcp::connect(address, formats::Bincode::default);
                conn.config_mut().max_frame_length(usize::MAX);
                Ok(tarpc::client::new(stub_config, conn.await?).spawn())
            },
            TransportCodec::Json => {
                // JSON codec using [serde_json](https://docs.rs/serde_json) crate.
                let mut conn = tcp::connect(address, formats::Json::default);
                conn.config_mut().max_frame_length(usize::MAX);
                Ok(tarpc::client::new(stub_config, conn.await?).spawn())
            },
            #[cfg(feature = "serde-transport-messagepack")]
            MessagePack => {
                /// MessagePack codec using [rmp-serde](https://docs.rs/rmp-serde) crate.
                let mut conn = tcp::connect(address, formats::MessagePack::default);
                conn.config_mut().max_frame_length(usize::MAX);
                Ok(tarpc::client::new(stub_config, conn.await?).spawn())
            },
            #[cfg(feature = "serde-transport-cbor")]
            Cbor => {
                /// CBOR codec using [serde_cbor](https://docs.rs/serde_cbor) crate.
                let mut conn = tcp::connect(address, formats::Cbor::default);
                conn.config_mut().max_frame_length(usize::MAX);
                Ok(tarpc::client::new(stub_config, conn.await?).spawn())
            },
        }
    }
    async fn reconnent(&mut self) -> Result<(), anyhow::Error> {
        self.channel = Self::new_channel(self.transport_codec.clone(), self.stub_config.clone(), self.address.as_str()).await?;
        self.is_shutdown.set(false);
        Ok(())
    }
}
