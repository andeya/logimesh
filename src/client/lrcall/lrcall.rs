// Copyright Andeya Lee 2024
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.
//! Local and remote Cclient.

use crate::client::balance::{LoadBalance, RpcChange};
use crate::client::channel::{RpcChannel, RpcConfig};
use crate::client::core::stub::Stub;
use crate::client::core::{Config, RpcError};
use crate::client::discover::{Discover, Discovery, Instance, InstanceCluster};
use crate::client::ClientError;
use crate::component::Component;
use crate::net::Address;
use crate::server::Serve;
use crate::transport::codec::Codec;
use futures_util::{select, FutureExt};
use std::collections::HashSet;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use tokio::sync::Notify;
use tracing::{trace, warn};

/// A full client stbu config.
#[non_exhaustive]
pub struct Builder<S, D, LB, RF>
where
    S: Serve + 'static,
    S::Req: crate::serde::Serialize + Send + 'static,
    S::Resp: for<'de> crate::serde::Deserialize<'de> + Send + 'static,
    D: Discover,
    LB: LoadBalance<S>,
    RF: Fn(&Result<S::Resp, RpcError>, u32) -> bool,
{
    /// The implementer of the local service.
    pub(crate) component: Component<S>,
    /// discover instance.
    pub(crate) discover: Arc<D>,
    /// load balance instance.
    pub(crate) load_balance: Arc<LB>,
    /// transport codec type.
    pub(crate) transport_codec: Codec,
    /// Settings that control the behavior of the underlying client.
    pub(crate) core_config: Config,
    /// A callback function for judging whether to re-initiate the request.
    pub(crate) retry_fn: Option<RF>,
}

/// A full client stbu config extend.
#[non_exhaustive]
pub struct ConfigExt {
    /// The number of requests that can be in flight at once.
    /// `max_in_flight_requests` controls the size of the map used by the client
    /// for storing pending requests.
    max_in_flight_requests: usize,
    /// The number of requests that can be buffered client-side before being sent.
    /// `pending_requests_buffer` controls the size of the channel clients use
    /// to communicate with the request dispatch task.
    pending_request_buffer: usize,
}

impl Default for ConfigExt {
    fn default() -> Self {
        let config: Config = Default::default();
        Self {
            max_in_flight_requests: config.max_in_flight_requests,
            pending_request_buffer: config.pending_request_buffer,
        }
    }
}

impl ConfigExt {
    /// The number of requests that can be in flight at once.
    /// `max_in_flight_requests` controls the size of the map used by the client
    /// for storing pending requests.
    /// Default is 1000.
    pub fn max_in_flight_requests(mut self, max_in_flight_requests: usize) -> Self {
        self.max_in_flight_requests = max_in_flight_requests;
        self
    }
    /// The number of requests that can be buffered client-side before being sent.
    /// `pending_requests_buffer` controls the size of the channel clients use
    /// to communicate with the request dispatch task.
    /// Default is 100.
    pub fn pending_request_buffer(mut self, pending_request_buffer: usize) -> Self {
        self.pending_request_buffer = pending_request_buffer;
        self
    }
}

impl<S, D, LB, RF> Builder<S, D, LB, RF>
where
    S: Serve + 'static,
    S::Req: crate::serde::Serialize + Send + 'static,
    S::Resp: for<'de> crate::serde::Deserialize<'de> + Send + 'static,
    D: Discover,
    LB: LoadBalance<S>,
    RF: Fn(&Result<S::Resp, RpcError>, u32) -> bool,
{
    /// Create a LRCall builder.
    pub fn new(component: Component<S>, discover: D, load_balance: LB) -> Self {
        Self {
            component,
            discover: Arc::new(discover),
            load_balance: Arc::new(load_balance),
            transport_codec: Default::default(),
            core_config: Default::default(),
            retry_fn: None,
        }
    }
    /// Set transport serde codec
    pub fn with_transport_codec(mut self, transport_codec: Codec) -> Self {
        self.transport_codec = transport_codec;
        self
    }
    /// The number of requests that can be in flight at once.
    /// `max_in_flight_requests` controls the size of the map used by the client
    /// for storing pending requests.
    /// Default is 1000.
    pub fn with_max_in_flight_requests(mut self, max_in_flight_requests: usize) -> Self {
        self.core_config.max_in_flight_requests = max_in_flight_requests;
        self
    }
    /// The number of requests that can be buffered client-side before being sent.
    /// `pending_requests_buffer` controls the size of the channel clients use
    /// to communicate with the request dispatch task.
    /// Default is 100.
    pub fn with_pending_request_buffer(mut self, pending_request_buffer: usize) -> Self {
        self.core_config.pending_request_buffer = pending_request_buffer;
        self
    }
    /// Set a callback function for judging whether to re-initiate the request.
    pub fn with_retry_fn(mut self, retry_fn: RF) -> Self {
        self.retry_fn = Some(retry_fn);
        self
    }
    /// Set some default extension configurations.
    pub fn with_config_ext(mut self, config_ext: ConfigExt) -> Self {
        self.core_config.max_in_flight_requests = config_ext.max_in_flight_requests;
        self.core_config.pending_request_buffer = config_ext.pending_request_buffer;
        self
    }
    /// Spawn a local and remote client.
    pub async fn try_spawn(self) -> Result<LRCall<S, D, LB, RF>, ClientError> {
        LRCall {
            config: self,
            notify: Arc::new(Notify::new()),
            use_rpc: Arc::new(AtomicBool::new(false)),
        }
        .warm_up()
        .await
    }
}

/// A local and remote client.
pub struct LRCall<S, D, LB, RF>
where
    S: Serve + 'static,
    S::Req: crate::serde::Serialize + Send + 'static,
    S::Resp: for<'de> crate::serde::Deserialize<'de> + Send + 'static,
    D: Discover,
    LB: LoadBalance<S>,
    RF: Fn(&Result<S::Resp, RpcError>, u32) -> bool,
{
    config: Builder<S, D, LB, RF>,
    notify: Arc<Notify>,
    use_rpc: Arc<AtomicBool>,
}

impl<S, D, LB, RF> LRCall<S, D, LB, RF>
where
    S: Serve + 'static,
    S::Req: crate::serde::Serialize + Send + 'static,
    S::Resp: for<'de> crate::serde::Deserialize<'de> + Send + 'static,
    D: Discover,
    LB: LoadBalance<S>,
    RF: Fn(&Result<S::Resp, RpcError>, u32) -> bool,
{
    async fn warm_up(self) -> Result<Self, ClientError> {
        let discovery = self.config.discover.discover(&self.config.component.endpoint).await?;
        let mut channels: Vec<RpcChannel<S>> = Vec::new();
        match discovery.instance_cluster {
            InstanceCluster::Lpc => self.use_rpc.store(false, Ordering::Release),
            InstanceCluster::Rpc(instances) => {
                for instance in instances {
                    self.use_rpc.store(true, Ordering::Release);
                    let channel = RpcChannel::new(RpcConfig {
                        instance,
                        transport_codec: self.config.transport_codec,
                        core_config: self.config.core_config.clone(),
                    })
                    .await;
                    match channel {
                        Ok(channel) => {
                            channels.push(channel);
                        },
                        Err(e) => {
                            warn!("[LOGIMESH] TCP connection establishment failed: {:?}", e)
                        },
                    }
                }
            },
        }
        let prev = channels.clone();
        self.config.load_balance.start_balance(channels);
        if let Some(mut recv_change) = self.config.discover.watch(None) {
            let load_balance = self.config.load_balance.clone();
            let transport_codec = self.config.transport_codec;
            let core_config = self.config.core_config.clone();
            let notify = self.notify.clone();
            let use_rpc = self.use_rpc.clone();
            tokio::spawn(async move {
                let mut prev = prev;
                loop {
                    select! {
                        _ = notify.notified().fuse() => {
                            return;
                        },
                        discovery = recv_change.recv().fuse() => match discovery {
                            Ok(Discovery{instance_cluster:InstanceCluster::Lpc,..}) => {
                                use_rpc.store(false, Ordering::Release);
                                prev.clear();
                            },
                            Ok(Discovery{instance_cluster:InstanceCluster::Rpc(next),..}) => {
                                use_rpc.store(true, Ordering::Release);
                                match Self::diff_and_dial(transport_codec, &core_config, &mut prev, next).await {
                                    Ok(changes) => {
                                        load_balance.rebalance(changes);
                                    },
                                    Err(err) => warn!("[LOGIMESH] TCP connection establishment failed: {:?}", err),
                                }
                            },
                            Err(err) => warn!("[LOGIMESH] discovering subscription error: {:?}", err),
                        },
                    }
                }
            });
        }
        Ok(self)
    }

    async fn diff_and_dial(transport_codec: Codec, core_config: &Config, prev: &mut Vec<RpcChannel<S>>, next: Vec<Arc<Instance>>) -> Result<Option<RpcChange<S>>, ClientError>
    where
        S: Serve,
    {
        let mut added: Vec<RpcChannel<S>> = Vec::new();
        let mut updated: Vec<RpcChannel<S>> = Vec::new();
        let mut removed: Vec<Address> = Vec::new();

        let mut next_set = HashSet::with_capacity(next.len());

        for i in &next {
            next_set.insert(i.address.clone());
        }

        for instance in &next {
            let mut is_new: bool = true;
            for c in prev.iter_mut() {
                if &instance.address == &c.config().instance.address {
                    is_new = false;
                    if &c.config().instance != instance {
                        let updated_channel = c.clone_update_instance(instance.clone());
                        *c = updated_channel.clone();
                        updated.push(updated_channel);
                    }
                    break;
                }
            }
            if is_new {
                let channel = RpcChannel::new(RpcConfig {
                    instance: instance.clone(),
                    transport_codec: transport_codec,
                    core_config: core_config.clone(),
                })
                .await;
                match channel {
                    Ok(channel) => {
                        added.push(channel);
                    },
                    Err(e) => {
                        warn!("[LOGIMESH] failed to connect: {e:?}");
                    },
                }
            }
        }

        for i in prev.iter() {
            if !next_set.contains(&i.config().instance.address) {
                removed.push(i.config().instance.address.clone());
            }
        }

        if removed.len() > 0 {
            prev.retain_mut(|c| {
                for address in &removed {
                    if &c.config().instance.address == address {
                        return false;
                    }
                }
                return true;
            });
        }

        prev.append(&mut added);

        let changed = !added.is_empty() || !removed.is_empty() || !updated.is_empty();
        if changed {
            Ok(Some(RpcChange {
                all: prev.clone(),
                added,
                updated,
                removed,
            }))
        } else {
            Ok(None)
        }
    }
}

impl<S, D, LB, RF> Stub for LRCall<S, D, LB, RF>
where
    S: Serve + Clone + 'static,
    S::Req: crate::serde::Serialize + Send + Clone + 'static,
    S::Resp: for<'de> crate::serde::Deserialize<'de> + Send + 'static,
    D: Discover,
    LB: LoadBalance<S>,
    RF: Fn(&Result<S::Resp, RpcError>, u32) -> bool,
{
    type Req = S::Req;

    type Resp = S::Resp;

    // TODO: Think about whether to fallback to LPC after RPC fails?
    async fn call(&self, ctx: crate::context::Context, request: Self::Req) -> Result<Self::Resp, RpcError> {
        let use_rpc = self.use_rpc.load(Ordering::Acquire);
        if let Some(retry_fn) = &self.config.retry_fn {
            if use_rpc {
                let mut picker = self.config.load_balance.get_picker();
                for i in 1.. {
                    if let Some(channel) = picker.next() {
                        let result = channel.call(ctx, request.clone()).await;
                        if let Err(RpcError::Shutdown) = result {
                            // TODO: Change to asynchronous processing
                            match channel.reconnent().await {
                                Ok(_) => trace!("[LOGIMESH] success to reconnect"),
                                Err(e) => warn!("[LOGIMESH] failed to reconnect: {e:?}"),
                            };
                        }
                        if (retry_fn)(&result, i) {
                            trace!("[LOGIMESH] Retrying on attempt {i}");
                            continue;
                        }
                        return result;
                    } else {
                        // When there is no connection, fallback to local call (LPC)
                        warn!("[LOGIMESH] As there is no connection, fallback to local call.");
                        return self.config.component.serve.call(ctx, request).await;
                    }
                }
                unreachable!("[LOGIMESH] Wow, that was a lot of attempts!");
            } else {
                for i in 1.. {
                    let result = self.config.component.serve.call(ctx, request.clone()).await;
                    if (retry_fn)(&result, i) {
                        trace!("[LOGIMESH] Retrying on attempt {i}");
                        continue;
                    }
                    return result;
                }
                unreachable!("[LOGIMESH] Wow, that was a lot of attempts!");
            }
        } else {
            if use_rpc {
                let mut picker = self.config.load_balance.get_picker();
                if let Some(channel) = picker.next() {
                    let result = channel.call(ctx, request).await;
                    if let Err(RpcError::Shutdown) = result {
                        // TODO: Change to asynchronous processing
                        match channel.reconnent().await {
                            Ok(_) => trace!("[LOGIMESH] success to reconnect"),
                            Err(e) => warn!("[LOGIMESH] failed to reconnect: {e:?}"),
                        };
                    }
                    return result;
                } else {
                    // When there is no connection, fallback to local call (LPC)
                    warn!("[LOGIMESH] As there is no connection, fallback to local call.");
                    return self.config.component.serve.call(ctx, request).await;
                }
            } else {
                return self.config.component.serve.call(ctx, request).await;
            }
        }
    }
}

impl<S, D, LB, RF> Drop for LRCall<S, D, LB, RF>
where
    S: Serve + 'static,
    S::Req: crate::serde::Serialize + Send + 'static,
    S::Resp: for<'de> crate::serde::Deserialize<'de> + Send + 'static,
    D: Discover,
    LB: LoadBalance<S>,
    RF: Fn(&Result<S::Resp, RpcError>, u32) -> bool,
{
    fn drop(&mut self) {
        self.notify.notify_waiters();
    }
}
