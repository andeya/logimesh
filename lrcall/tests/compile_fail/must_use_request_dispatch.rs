use lrcall::client;

#[lrcall::service]
trait World {
    async fn hello(name: String) -> String;
}

fn main() {
    let (client_transport, _) = lrcall::transport::channel::unbounded();

    #[deny(unused_must_use)]
    {
        WorldClient::<HelloService>::rpc_client((client::Config::default(), client_transport).into()).dispatch;
    }
}
