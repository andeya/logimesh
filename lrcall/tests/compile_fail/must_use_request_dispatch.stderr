error: unused `RequestDispatch` that must be used
  --> tests/compile_fail/must_use_request_dispatch.rs:13:9
   |
13 |         WorldClient::<HelloService>::rpc_client((client::Config::default(), client_transport).into()).dispatch;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> tests/compile_fail/must_use_request_dispatch.rs:11:12
   |
11 |     #[deny(unused_must_use)]
   |            ^^^^^^^^^^^^^^^
help: use `let _ = ...` to ignore the resulting value
   |
13 |         let _ = WorldClient::<HelloService>::rpc_client((client::Config::default(), client_transport).into()).dispatch;
   |         +++++++
